File: python/8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python3
#
# [8] 字符串转换整数 (atoi)
#
# @lc code=start
INT_MAX = (2**31 - 1)
INT_MIN = (-2**31)


class Automaton:
    # 状态机
    def __init__(self) -> None:
        self.ans = 0
        self.sign = 1
        self.state = 'start'
        self.table = {
            'start': ['start', 'signed', 'in_number', 'end'],
            'signed': ['end', 'end', 'in_number', 'end'],
            'in_number': ['end', 'end', 'in_number', 'end'],
            'end': ['end', 'end', 'end', 'end']
        }

    def get_col(self, s: str) -> int:
        if s.isspace():
            return 0
        elif s == '+' or s == '-':
            return 1
        elif s.isdigit():
            return 2
        else:
            return 3

    def get(self, s: str):
        self.state = self.table[self.state][self.get_col(s)]
        if self.state == 'signed':
            self.sign = -1 if s == '-' else 1
        elif self.state == 'in_number':
            self.ans = self.ans * 10 + int(s)
            self.ans = min(self.ans, -INT_MIN) if self.sign == -1 else min(
                self.ans, INT_MAX)


class Solution:
    # 状态机
    def myAtoi(self, s: str) -> int:
        automaton = Automaton()
        for ch in s:
            automaton.get(ch)
        return automaton.ans * automaton.sign

    # 非状态机
    def myAtoi1(self, s: str) -> int:
        s = s.lstrip()
        isNeg = False
        if not s:
            return 0
        elif (s[0] == '+' or s[0] == '-'):
            isNeg = True if s[0] == '-' else False
            s = s[1:]
        rec = 0

        for i, ch in enumerate(s):
            if ch.isnumeric():
                rec = rec * 10 + ord(ch) - ord('0')
            else:
                break
        rec = rec if not isNeg else -rec
        if rec < INT_MIN:
            return INT_MIN
        elif rec > INT_MAX:
            return INT_MAX
        return rec


print(Solution().myAtoi("   -42"))
# @lc code=end

File: cpp/8.字符串转换整数-atoi.cpp
/*
 * @lc app=leetcode.cn id=8 lang=cpp
 *
 * [8] 字符串转换整数 (atoi)
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Automaton
{
    string state = "start";

    unordered_map<string, vector<string>> table = {
        {"start", {"start", "signed", "number", "end"}},
        {"signed", {"end", "end", "number", "end"}},
        {"number", {"end", "end", "number", "end"}},
        {"end", {"end", "end", "end", "end"}}};

    int get_col(char ch)
    {
        if (isspace(ch)) {
            return 0;
        }
        else if (ch == '+' || ch == '-') {
            return 1;
        }
        else if (isdigit(ch)) {
            return 2;
        }
        else {
            return 3;
        }
    }

public:
    int sign = 1;

    long long int ans = 0;

    void get(char ch)
    {
        state = table[state][get_col(ch)];

        if (state == "number") {
            ans = ans * 10 + ch - '0';
            // 这里不能改动ans为-ans，因为最终的答案要由automaton.ans * automaton.sign计算得出。
            // 把所有输入都当成正数来处理，方便逻辑统一。
            ans = sign == 1 ? min(ans, (long long)INT_MAX)
                            : min(ans, -(long long)INT_MIN);
        }
        else if (state == "signed") {
            sign = ch == '+' ? 1 : -1;
        }
    }
};


class Solution
{
public:
    // 自动机
    int myAtoi(string s)
    {
        Automaton automaton;
        for (char ch : s) {
            automaton.get(ch);
        }
        return automaton.ans * automaton.sign;
    }
};
// @lc code=end
int main()
{
    cout << (Solution()).myAtoi("   -42") << endl;
}